src/crypto/nss.rs

```diff
@@ -87,123 +96,435 @@ pub(crate) fn serialize_key(_curve: ECDSACurve, key: &[u8]) -> Result<(ByteBuf,
+    spki.push(0x03);
+    spki.extend_from_slice(spk_len);
+    spki.push(0x0); // no trailing zeros
+    spki.extend_from_slice(&cose_key_sec1);
+
+    Ok(spki)
 }
 
 /// This is run by the platform when starting a series of transactions with a specific authenticator.
-// pub(crate) fn initialize() {
-//     unimplemented!()
-// }
+//pub(crate) fn initialize() { }
 
 /// Generates an encapsulation for the authenticator’s public key and returns the message
 /// to transmit and the shared secret.
-pub(crate) fn encapsulate(key: &COSEKey) -> Result<ECDHSecret> {
-    if let COSEKeyType::EC2(ec2key) = &key.key {
-        let alg = to_nss_alg(key.alg)?;
-        let keypair: KeyPair<Ephemeral> = KeyPair::generate(alg)?;
-        let (private_key, public_key) = keypair.split();
-        encapsulate_helper(ec2key, key.alg, &public_key.to_bytes()?, private_key)
-    } else {
-        Err(BackendError::UnsupportedKeyType)
-    }
-}
+///
+/// `key` is the authenticator's (peer's) public key.
+pub(crate) fn encapsulate(peer_cose_key: &COSEKey) -> Result<ECDHSecret> {
+    // Generate an ephmeral keypair to do ECDH with the authenticator.
+    // This is "platformKeyAgreementKey".
+    let ec2key = match peer_cose_key.key {
+        COSEKeyType::EC2(ref ec2key) => ec2key,
+        _ => return Err(BackendError::UnsupportedKeyType),
+    };
 
-fn encapsulate_helper(
-    key: &COSEEC2Key,
-    alg: COSEAlgorithm,
-    public_key: &[u8],
-    private_key: PrivateKey<Ephemeral>,
-) -> Result<ECDHSecret> {
-    let (x, y) = serialize_key(key.curve, public_key)?;
-    let my_public_key = COSEKey {
-        alg,
-        key: COSEKeyType::EC2(COSEEC2Key {
-            curve: key.curve.clone(),
-            x: x.to_vec(),
-            y: y.to_vec(),
-        }),
+    let mut oid = match ec2key.curve {
+        ECDSACurve::SECP256R1 => SECItemBorrowed::wrap(DER_OID_P256_BYTES),
+        ECDSACurve::SECP384R1 => SECItemBorrowed::wrap(DER_OID_P384_BYTES),
+        ECDSACurve::SECP521R1 => SECItemBorrowed::wrap(DER_OID_P521_BYTES),
+        x => return Err(BackendError::UnsupportedCurve(x)),
     };
-    let key_bytes = parse_key(key.curve, &key.x, &key.y)?;
-    let peer_public_key = UnparsedPublicKey::new(private_key.algorithm(), &key_bytes);
+    let oid_ptr: *mut SECItem = oid.as_mut();
 
-    let shared_secret = private_key.agree(&peer_public_key)?;
-    let digest = shared_secret.derive(|input| digest(&HashAlgorithm::SHA256, input))?;
+    let slot = Slot::internal()?;
+
+    let mut client_public_ptr = ptr::null_mut();
+    let client_private = unsafe {
+        // Type of `param` argument depends on mechanism. For EC keygen it is
+        // `SECKEYECParams *` which is a typedef for `SECItem *`.
+        PK11_GenerateKeyPairWithOpFlags(
+            *slot,
+            CKM_EC_KEY_PAIR_GEN,
+            oid_ptr.cast(),
+            &mut client_public_ptr,
+            PK11_ATTR_SESSION | PK11_ATTR_INSENSITIVE | PK11_ATTR_PUBLIC,
```

Comment:
This will probably fail, if Firefox is running in FIPS-mode, as the creation of INSENSITIVE keys are not allowed under FIPS, AFAIK. Just a comment, as I know, this is not a focus of Firefox.

Reply:
Good catch, I think we can omit the flags and use the token defaults.